### PR TITLE
Implement assets entity

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
### Overview:
This pull request finalizes several key interfaces and introduces robust asset management functionality for both `MemoryLedger` and `TextLedger`, focusing on accessor, mutator, and mirroring methods. It also includes updates to precision handling, asset standardization, and test coverage.

### Key Changes:

1. **Completion of Interfaces**:
   - Finalized the implementation of asset-related interfaces across both `MemoryLedger` and `TextLedger`.
   - Ensured consistency in schema enforcement and method signatures for handling assets in both ledger systems.

2. **Implementation of Mirroring Logic and Introduction of `sanitize_assets` Method**:
   - Implemented the `mirror_assets` method to synchronize the current ledger state with a target state. The method adds missing assets, updates existing ones, and (optionally) deletes assets that are no longer present in the target.
   - Introduced a `sanitize_assets` method to ensure that asset data adheres to the required schema before mirroring operations. This improves data integrity and consistency across operations.

3. **Implementing Accessor and Mutator Methods for `MemoryLedger`**:
   - Added accessor and mutator methods (`add_asset`, `modify_asset`, and `delete_asset`) for managing assets in `MemoryLedger`.
   - Each method ensures correct handling of assets uniquely identified by `ticker` and `date`, with proper error handling for duplicates or non-existent assets.

4. **Adding Base Tests for Assets**:
   - Added base tests covering accessor, mutator, and mirroring methods for assets. These tests ensure that the logic is thoroughly validated and works consistently in both `MemoryLedger` and `TextLedger`.
   - Included test cases for handling precision and date-related edge cases (such as assets with `None` dates).

5. **Updating Usage of Precision**:
   - Updated the `precision` method to correctly retrieve the `increment` value for assets based on the `ticker` and `date`.
   - Ensured that `None` dates are properly handled by treating them as `pd.NaT`, making it easier to compare and manipulate assets with or without specific dates.

6. **Implementing Accessor and Mutator Methods for `TextLedger`**:
   - Implemented accessor and mutator methods for managing assets in `TextLedger`, including reading and writing assets to CSV files.
   - Methods such as `add_asset`, `modify_asset`, and `delete_asset` now work with `TextLedger`, ensuring assets are correctly read from and written to `assets.csv` files in a fixed-width format suitable for version control.

### Review Recommendation:
It is recommended to review the changes **commit by commit** to understand the incremental updates and ensure proper context for each stage of implementation.

### CODE REVIEW

File: pyledger/ledger_engine.py [hunk 1]
```diff
@@ -117,9 +105,105 @@ def standardize_revaluations(cls, df: pd.DataFrame,
...
         return df
 
+    @abstractmethod
+    def revaluations(self) -> pd.DataFrame:
+        """Retrieves all revaluation entries.
+
+        Returns:
+            pd.DataFrame: DataFrame with columns `date` (datetime64[ns]), `account` (str),
+                          `credit` (Int64), `debit` (Int64, optional), `description` (str),
+                          and `price` (Float64, optional).
```
Comment (lasuk) [inline pyledger/ledger_engine.py:118]: I would not list the columns:
```
Returns:
    pd.DataFrame with columns specified in REVALUATION_SCHEMA.
```
Instead of:
```
        Returns:
            pd.DataFrame: DataFrame with columns `date` (datetime64[ns]), `account` (str),
                          `credit` (Int64), `debit` (Int64, optional), `description` (str),
                          and `price` (Float64, optional).
```

File: pyledger/ledger_engine.py [hunk 2]
```diff
@@ -117,9 +105,105 @@ def standardize_revaluations(cls, df: pd.DataFrame,
...
+        account: str,
+        credit: int,
+        description: str,
+        debit: int = None,
+        price: float = None
+    ) -> None:
+        """Adds a new revaluation.
+
+        The unique identifier is a combination of `date` and `account`.
+        If `credit` is None, it will be assigned the same value as `debit`.
```
Comment (lasuk) [inline pyledger/ledger_engine.py:134]: This line is redundant and can be dropped (also appears in argument description).

~~If `credit` is None, it will be assigned the same value as `debit`.~~

(Also applies to modify_revaluation)

File: pyledger/ledger_engine.py [hunk 3]
```diff
@@ -117,9 +105,105 @@ def standardize_revaluations(cls, df: pd.DataFrame,
...
+
+        Args:
+            date (datetime.date): Date of the revaluation.
+            account (str): Account associated with the revaluation.
```
Comment (lasuk) [inline pyledger/ledger_engine.py:138]:
```
account (str): Account (e.g. "1020") or range of accounts (e.g. "1020:1099") to revalue.
```
instead of:
```
account (str): Account associated with the revaluation.
```

(Also applies to modify_revaluation)

File: pyledger/ledger_engine.py [hunk 4]
```diff
@@ -117,9 +105,105 @@ def standardize_revaluations(cls, df: pd.DataFrame,
...
+        dates: List[datetime.date],
+        allow_missing: bool = False
+    ) -> None:
+        """Removes revaluations.
+
+        The unique identifier is a combination of `date` and `account`.
+
+        Args:
+            accounts (List[str]): List of accounts associated with the revaluations to remove.
+            dates (datetime.date): List of dates corresponding to the revaluations to remove.
```
Comment (lasuk) [inline pyledger/ledger_engine.py:184]: Do accounts and dates need to be lists of same length?

How about using a list of tuples?
```
Args:
    id (List[Tuple[str, datetime.date | None]): List of accounts and dates of revaluations to remove.
```
Or providing the date as a singular value?
```
Args:
    accounts (List[str]): List of accounts of revaluations to remove.
    date (datetime.date | None): The date of the revaluations to remove.
```
Or providing the date and the account as a singular values?
```
Args:
    accounts (str): Account or account range of revaluations to remove.
    date (datetime.date | None): The date of the revaluations to remove.
```

Instead of:
```
Args:
    accounts (List[str]): List of accounts associated with the revaluations to remove.
    dates (datetime.date): List of dates corresponding to the revaluations to remove.
```

File: pyledger/ledger_engine.py [hunk 5]
```diff
@@ -117,9 +105,105 @@ def standardize_revaluations(cls, df: pd.DataFrame,
...
+            debit (int, optional): Debit amount for the revaluation.
+            description (str): Description of the revaluation.
+            price (float, optional): Price associated with the revaluation.
+        """
+
+    @abstractmethod
+    def delete_revaluations(
+        self,
+        accounts: List[str],
+        dates: List[datetime.date],
```
Comment (lasuk) [inline pyledger/ledger_engine.py:175]: Align type hint and doc string.

File: pyledger/ledger_engine.py [hunk 6]
```diff
@@ -117,9 +105,105 @@ def standardize_revaluations(cls, df: pd.DataFrame,
...
+            allow_missing (bool, optional): If True, no error is raised if a revaluation entry is
+                                            not found. Defaults to False.
+        """
+
+    def mirror_revaluations(self, target: pd.DataFrame, delete: bool = False) -> dict:
+        """Aligns revaluations with a desired target state.
+
+        Args:
+            target (pd.DataFrame): DataFrame with revaluations in the same format as defined
+                                   by the schema.
```
Comment (lasuk) [inline pyledger/ledger_engine.py:194]:
```
target (pd.DataFrame): DataFrame adhering to REVALUATION_SCHEMA.
```
instead of:
```
target (pd.DataFrame): DataFrame with revaluations in the same format as defined
                       by the schema.
```

File: pyledger/ledger_engine.py [hunk 7]
```diff
@@ -171,6 +255,7 @@ def dump_to_zip(self, archive_path: str):
             archive.writestr('ledger.csv', self.ledger().to_csv(index=False))
             archive.writestr('tax_codes.csv', self.tax_codes().to_csv(index=False))
             archive.writestr('accounts.csv', self.accounts().to_csv(index=False))
+            archive.writestr('assets.csv', self.assets().to_csv(index=False))
```
Comment (lasuk) [inline pyledger/ledger_engine.py:258]: Please also implement revaluations and price in 'dump_to_zip', 'restore', 'restore_from_zip' and 'clear'.

File: pyledger/ledger_engine.py [hunk 8]
```diff
@@ -1289,6 +1383,21 @@ def add_price(
                                         date, and currency if one exists. Defaults to False.
         """
 
+    @abstractmethod
+    def modify_price(
+        self, ticker: str, date: datetime.date, currency: str, price: float, overwrite: bool = False
+    ) -> None:
+        """Modifies a price within the price history.
```
Comment (lasuk) [inline pyledger/ledger_engine.py:1390]:
```
Modifies an observation in the price history.
```
Instead of:
```
Modifies a price within the price history.
```

File: pyledger/ledger_engine.py [hunk 9]
```diff
@@ -1289,6 +1383,21 @@ def add_price(
...
+
+        Args:
+            ticker (str): Asset identifier.
+            date (datetime.date): Date on which the price is recorded.
+            currency (str): Currency in which the price is quoted.
+            price (float): Value of the asset as of the given date.
+            overwrite (bool, optional): Overwrite an existing price definition with the same ticker,
+                                        date, and currency if one exists. Defaults to False.
```
Comment (lasuk) [inline pyledger/ledger_engine.py:1398]: Why do we need an argument 'overwrite'?

File: pyledger/ledger_engine.py [hunk 10]
```diff
@@ -1347,3 +1445,170 @@ def standardize_prices(cls, df: pd.DataFrame) -> pd.DataFrame:
...
+            pd.DataFrame: DataFrame with columns `ticker` (str), `increment` (float),
+                          and `date` (datetime64[ns], optional).
+        """
+
+    @abstractmethod
+    def add_asset(
+        self,
+        ticker: str,
+        increment: float,
+        date: pd.Timestamp
```
Comment (lasuk) [inline pyledger/ledger_engine.py:1512]: I believe 'date' is optional

(Also applies to modify_asset and delete_asset)

File: pyledger/memory_ledger.py [hunk 11]
```diff
@@ -27,6 +26,7 @@ def __init__(self, reporting_currency: str = "USD") -> None:
         self._prices = self.standardize_prices(None)
         self._tax_codes = self.standardize_tax_codes(None)
         self._accounts = self.standardize_accounts(None)
+        self._assets = self.standardize_assets(None)
```
Comment (lasuk) [inline pyledger/memory_ledger.py:29]: Also add revaluations.

File: pyledger/standalone_ledger.py [hunk 12]
```diff
@@ -485,11 +489,25 @@ def price(
         return (currency, prc.iloc[-1].item())
 
     def precision(self, ticker: str, date: datetime.date = None) -> float:
-        return self.settings["precision"][ticker]
+        if ticker == "reporting_currency":
+            ticker = self.reporting_currency
+        date = pd.Timestamp(date) if date is not None else pd.NaT
```
Comment (lasuk) [inline pyledger/standalone_ledger.py:494]: If date is None, use the current date.

File: pyledger/standalone_ledger.py [hunk 13]
```diff
@@ -485,11 +489,25 @@ def price(
...
+        mask = (
+            (self._assets["ticker"] == ticker)
+            & ((self._assets["date"] == date) | pd.isna(self._assets["date"]))
+        )
```
Comment (lasuk) [inline pyledger/standalone_ledger.py:498]: If entry with the specified date does not exist for the given ticke, 
we need to look up the last update for the given ticker prior to `date`.

File: pyledger/tests/base_test.py [hunk 14]
```diff
@@ -93,3 +107,6 @@ class BaseTest(ABC):
     LEDGER_ENTRIES = LedgerEngine.standardize_ledger_columns(pd.read_csv(
         StringIO(STRIPPED_CSV), skipinitialspace=True, comment="#", skip_blank_lines=True
     ))
+    ASSETS = LedgerEngine.standardize_assets(
+        pd.read_csv(StringIO(ASSETS_CSV), skipinitialspace=True)
```
Comment (lasuk) [inline pyledger/tests/base_test.py:111]: Call standardize_assets.

File: pyledger/tests/base_test.py [hunk 15]
```diff
@@ -79,6 +79,20 @@
 
 # flake8: enable
 
+ASSETS_CSV = """
```
Comment (lasuk) [inline pyledger/tests/base_test.py:82]: Test data should contain both entries with and without date

File: pyledger/tests/base_test_assets.py
Comment (lasuk) [inline pyledger/tests/base_test_assets.py:1]: We need to test somewhere, whether precision() finds the correct increment.

**Review by lasuk — COMMENTED** [review]
Thanks for implementing 'asset' accusers, mutators and mirroring.

Please find my comments below.

File: pyledger/constants.py [hunk 16]
```diff
@@ -58,20 +59,27 @@
...
+       CHF,      0.01,
+       EUR,      0.01,
+       GBP,      0.01,
+       JPY,      1.00,
+       NZD,      0.01,
+       NOK,      0.01,
+       SEK,      0.01,
+       USD,      0.01,
+"""
+DEFAULT_ASSETS = pd.read_csv(StringIO(ASSETS_CSV), skipinitialspace=True)
```
[the remark this answers is not in the document]
Comment (AlexTheWizardL) [reply pyledger/constants.py:83]: We can not import `LedgerEngine` here due to a circular import:
```terminal
______________________________________________________________________________________________________________________ ERROR collecting pyledger/test_ledger.py _______________________________________________________________________________________________________________________
ImportError while importing test module '/pyledger/pyledger/test_ledger.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pyledger/__init__.py:22: in <module>
    from .ledger_engine import LedgerEngine
pyledger/ledger_engine.py:16: in <module>
    from .constants import (
pyledger/constants.py:5: in <module>
    from pyledger import LedgerEngine
E   ImportError: cannot import name 'LedgerEngine' from partially initialized module 'pyledger' (most likely due to a circular import) (/pyledger/pyledger/__init__.py)
```

I can just hit to the `enforce_schema()` instead. What do you think?

File: pyledger/ledger_engine.py (hunk 4 above)
Comment (AlexTheWizardL) [reply pyledger/ledger_engine.py:184]: I like to go with the idea of a singular values

File: pyledger/ledger_engine.py (hunk 7 above)
Comment (AlexTheWizardL) [reply pyledger/ledger_engine.py:258]: Will be done in the next PR, TODOs already saying that.

File: pyledger/ledger_engine.py (hunk 9 above)
Comment (AlexTheWizardL) [reply pyledger/ledger_engine.py:1398]: 'overwrite' is the wrong guy here, will remove it, thanks

File: pyledger/tests/base_test.py (hunk 14 above)
Comment (AlexTheWizardL) [reply pyledger/tests/base_test.py:111]: It is already there, or do I misunderstand?

Comment (lasuk) [reply pyledger/tests/base_test.py:111]: Thanks, my mistake.

**Review by lasuk — COMMENTED** [review]
Thanks for the changes.
- I've pushed four commits with mostly cosmetic changes with the aim to lay a very clear and concise ground with assets as a role model for further entities. Please review and shout if you disagree.
- Added a few comments.

File: pyledger/standalone_ledger.py [hunk 17]
```diff
@@ -484,12 +491,52 @@ def price(
...
+        elif not isinstance(date, datetime.date):
+            date = pd.to_datetime(date).date()
+
+        increment = self.assets_as_dict_of_df.get(ticker)
+        if increment is None:
+            raise ValueError(f"No data available for ticker '{ticker}'.")
+
+        increments = increment.loc[
+            (increment["date"].isna())
+            | (increment["date"].dt.normalize() <= pd.Timestamp(date)), "increment"
```
[the remark this answers is not in the document]
Comment (AlexTheWizardL) [reply pyledger/standalone_ledger.py:526]: Thanks for your clarification; here is what I’ve found out:

**Problem Explanation**  
Using `mask.idxmax()` returned the first occurrence of `True`, not the last, because `idxmax()` is designed to find the first maximum (or `True`) in a series. This is why it pointed to index `0` instead of the last `True` index, which was `1`.

```python
increment.loc[mask.idxmax(), 'increment']
```

**New Approach**  
To get the last `True` index in `mask` and update it in `increment`, use this approach:

```python
# Update the last True value in increment
increment.loc[mask[mask].index[-1], "increment"] = <new_value>
```

This approach works by filtering `mask` for `True` values (`mask[mask]`), then using `.index[-1]` to get the last `True` index, ensuring we update the correct row in `increment`.

**Comment by lasuk** [comment]
Merged. Thanks for the good foundation for accessors and mutators of future entities ...
